[JPEGXL] [Cocoa] Hook up color conversion to downlevel JPEG XL infrastructure

litherum opened the pull request:
#### 19b823079b6db3f9210dcc7e4034e0c2d317b92a
<pre>
[JPEGXL] [Cocoa] Hook up color conversion to downlevel JPEG XL infrastructure
<a href="https://bugs.webkit.org/show_bug.cgi?id=258267">https://bugs.webkit.org/show_bug.cgi?id=258267</a>
rdar://111012276

Reviewed by Cameron McCormack and Don Olmstead.

The existing JPEG XL infrastructure on downlevel builds uses the &quot;little cms&quot; library
from <a href="https://github.com/LuaDist/lcms/tree/master">https://github.com/LuaDist/lcms/tree/master</a> to perform color conversions. However,
on Cocoa platforms, we already have color conversion support in the platform. This
patch hooks up the existing colorspace conversion functionality in
PixelBufferConversion.cpp instead of the little cms library.

The existing infrastructure only works with RGBA8 sRGB images, so configuring all
the converters is actually pretty straightforward.

This patch isn&apos;t testable, because AppleJPEGXL will only be present in builds produced
by Apple&apos;s internal build system, but our testers use local builds to test with. I had
to verify manually that this patch is working correctly.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/VImageConverterPtr.h: Added.
(VImageConverterRefDerefTraits::refIfNotNull):
(VImageConverterRefDerefTraits::derefIfNotNull):
(adoptVImageConverter):
* Source/WebCore/PAL/pal/spi/cocoa/AppleJPEGXL/AppleJPEGXLSPI.h:
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::clear):
(WebCore::JPEGXLImageDecoder::processInput):
(WebCore::JPEGXLImageDecoder::imageOut):
(WebCore::JPEGXLImageDecoder::maybePerformColorSpaceConversion):
(WebCore::JPEGXLImageDecoder::clearColorTransform):
(WebCore::JPEGXLImageDecoder::prepareColorTransform):
(WebCore::JPEGXLImageDecoder::tryDecodeICCColorProfile):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.h:
* WebKitLibraries/DownlevelFrameworkStubs/120000/AppleJPEGXL.framework/AppleJPEGXL.tbd:
* WebKitLibraries/DownlevelFrameworkStubs/130000/AppleJPEGXL.framework/AppleJPEGXL.tbd:

Canonical link: <a href="https://commits.webkit.org/265440@main">https://commits.webkit.org/265440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec3972bf7093ed4d8ccd1c24d5923ff8ab1d1e51

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11112 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12546 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/10422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10915 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11093 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13340 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11058 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11962 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9181 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12950 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9255 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9843 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17077 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10326 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9995 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13232 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10441 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9613 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2616 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13883 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/10301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/PAL/pal/cocoa/VImageConverterPtr.h:
```diff
@@ -0,0 +1,51 @@
+inline VImageConverterPtr adoptVImageConverter(vImageConverterRef converter)
+{
+    return adoptRef<vImageConverter, WTF::RawPtrTraits<vImageConverter>, VImageConverterRefDerefTraits>(converter);
+}
```
There seems to be two different ways of dealing with `RefPtr` traits in the codebase.

One is like this, where we define a new type alias for refcounted pointers of this specific type, so that users do not need to type out the full `RefPtr<...>` with the traits type argument. A new `adoptXxx` function is needed for this type.

The other is to specialize `DefaultRefDerefTraits`, which I see done for some Gtk types (in RefPtrCairo.h and RefPtrFontconfig.h). With this, users can write `RefPtr<TheType>` and use `adoptRef`.

Is there a reason to choose one way over the other? We should probably have some guidance in Introduction.md in the `RefPtr` section. (Not something to do in this patch.)

----

But specifically for `vImageConverterPtr`: is this needed, or does `RetainPtr` work on it? I see in PixelBufferConversion.cpp that we use `adoptCF` on the return value of `vImageConverter_CreateWithCGImageFormat`.

heycam commented on Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
```diff
@@ -437,7 +438,117 @@ LCMSProfilePtr JPEGXLImageDecoder::tryDecodeICCColorProfile()
+    m_imageConverter = nullptr;
+}
+
+void JPEGXLImageDecoder::prepareColorTransform()
```
Is there any way we can use PixelBufferConverson.h/cpp's functionality to do the work of `prepareColorTransform` and `maybePerformColorSpaceConversion`? `convertImagePixels` is not set up to do a two stage "create the transform" then "apply the transform" conversion, but maybe we can just store the `CGColorSpaceRef` for the embedded ICC profile rather than the `vImageConverterPtr` when we get `JXL_DEC_COLOR_ENCODING`?

litherum replied:
There is! Updating the patch now.

heycam reviewed (commented):
You have two step 4s in the commit message.

unicodon reviewed (commented):
Looks good to me regarding non CG part 👍

donny-dont commented on Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
```diff
@@ -393,13 +396,11 @@ void JPEGXLImageDecoder::imageOut(size_t x, size_t y, size_t numPixels, const ui
 
-#if USE(LCMS)
-    if (m_iccTransform)
-        cmsDoTransform(m_iccTransform.get(), row, row, numPixels);
-#endif
+    maybePerformColorSpaceConversion(row, row, numPixels);
```
I don't see a default implementation of `maybePerformColorSpaceConversion`.

donny-dont reviewed (approved):
Approving the non-Apple parts with @unicodon

heycam commented on Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
```diff
@@ -242,7 +242,7 @@ void convertImagePixels(const ConstPixelBufferConversionView& source, const Pixe
 
 #if USE(ACCELERATE) && USE(CG)
-    if (source.format.alphaFormat == destination.format.alphaFormat && source.format.pixelFormat == destination.format.pixelFormat) {
+    if (source.format.alphaFormat == destination.format.alphaFormat && source.format.pixelFormat == destination.format.pixelFormat && source.format.colorSpace == destination.format.colorSpace) {
```
Hmm is this an existing bug?

litherum replied:
Yep. It seems I was the first person to trip over it.

heycam commented on Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
```diff
@@ -437,7 +437,82 @@ LCMSProfilePtr JPEGXLImageDecoder::tryDecodeICCColorProfile()
+{
+    size_t profileSize;
+    if (JxlDecoderGetICCProfileSize(m_decoder.get(), &s_pixelFormat, JXL_COLOR_PROFILE_TARGET_DATA, &profileSize) != JXL_DEC_SUCCESS)
+        return nullptr;
```
Here we are relying on libjxl always correctly writing to `profileSize` if `JXL_DEC_SUCCESS` is returned. Probably the chance of there being a bug here is low, but incorrect buffer sizes are a great source of security bugs. Can we initialize this to zero just to avoid any potential issue?

litherum replied:
Good idea.

heycam commented on Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
```diff
@@ -437,7 +437,82 @@ LCMSProfilePtr JPEGXLImageDecoder::tryDecodeICCColorProfile()
+    if (JxlDecoderGetColorAsICCProfile(m_decoder.get(), &s_pixelFormat, JXL_COLOR_PROFILE_TARGET_DATA, profileData.data(), profileData.size()) != JXL_DEC_SUCCESS)
+        return nullptr;
+
+    return adoptCF(CGColorSpaceCreateWithICCData(adoptCF(CFDataCreate(kCFAllocatorDefault, profileData.data(), profileSize)).get()));
```
Any reason not to create the `CFData` buffer first, and use that as the target for `JxlDecoderGetColorAsICCProfile`, rather than targeting the temporary vector buffer and copying from it?

litherum replied:
Yeah, I can use CFMutableData instead. Good idea.

heycam commented on Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
```diff
@@ -437,7 +437,82 @@ LCMSProfilePtr JPEGXLImageDecoder::tryDecodeICCColorProfile()
+    // https://developer.apple.com/documentation/accelerate/1399134-vimageconvert_anytoany?language=objc
+    // "The destination buffer may only alias the srcs buffers if vImageConverter_MustOperateOutOfPlace returns 0 and the respective scanlines of the aliasing buffers start at the same address."
+    // convertImagePixels() doesn't have any logic for this, so we can just pessimize here and assume aliasing is illegal.
+    std::unique_ptr<uint8_t[]> intermediateBuffer(new uint8_t[4  * numberOfPixels]);
```
Nit: one too many spaces after "4".

litherum replied:
😅

heycam commented on Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
```diff
@@ -437,7 +437,82 @@ LCMSProfilePtr JPEGXLImageDecoder::tryDecodeICCColorProfile()
+        .bytesPerRow = static_cast<unsigned>(4 * size().width()),
+        .rows = intermediateBuffer.get(),
+    };
+    convertImagePixels(source, destination, { static_cast<int>(numberOfPixels), 1 });
```
Feels a little odd to be pretending this is a _large_number_ x 1 dimension image. `bytesPerRow` is set to the right thing. I think we should just pass in `size()` here. Unless you are concerned about a mismatch between `size()` and `numberOfPixels`? (If so, you could validate that they match.)

litherum replied:
We can't - that's not the design of AppleJPEGXL. This code is all called from a callback that we give to AppleJPEGXL, and AppleJPEGXL will call the callback whenever it wants to emit a bunch of pixels in a row. We don't have the whole image here, we only have whatever chunk AppleJPEGXL decides to give us (which is guaranteed to be one-dimensional)

heycam replied:
OK. Kind of makes me feel like there should be a `convertImagePixels` overload that takes a number of pixels, instead of using `IntSize`. Can you confirm that `convertImagePixels` will do the right thing in terms of stride when we pass in an `IntSize` whose width does not correspond to the `bytesPerRow` we specify?

litherum committed:
[JPEGXL] [Cocoa] Hook up color conversion to downlevel JPEG XL infrastructure
https://bugs.webkit.org/show_bug.cgi?id=258267
rdar://111012276

Reviewed by Cameron McCormack and Don Olmstead.

The existing JPEG XL infrastructure on downlevel builds uses the "little cms" library
from https://github.com/LuaDist/lcms/tree/master to perform color conversions. However,
on Cocoa platforms, we already have color conversion support in the platform. This
patch hooks up the existing colorspace conversion functionality in
PixelBufferConversion.cpp instead of the little cms library.

The existing infrastructure only works with RGBA8 sRGB images, so configuring all
the converters is actually pretty straightforward.

This patch isn't testable, because AppleJPEGXL will only be present in builds produced
by Apple's internal build system, but our testers use local builds to test with. I had
to verify manually that this patch is working correctly.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/VImageConverterPtr.h: Added.
(VImageConverterRefDerefTraits::refIfNotNull):
(VImageConverterRefDerefTraits::derefIfNotNull):
(adoptVImageConverter):
* Source/WebCore/PAL/pal/spi/cocoa/AppleJPEGXL/AppleJPEGXLSPI.h:
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::clear):
(WebCore::JPEGXLImageDecoder::processInput):
(WebCore::JPEGXLImageDecoder::imageOut):
(WebCore::JPEGXLImageDecoder::maybePerformColorSpaceConversion):
(WebCore::JPEGXLImageDecoder::clearColorTransform):
(WebCore::JPEGXLImageDecoder::prepareColorTransform):
(WebCore::JPEGXLImageDecoder::tryDecodeICCColorProfile):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.h:
* WebKitLibraries/DownlevelFrameworkStubs/120000/AppleJPEGXL.framework/AppleJPEGXL.tbd:
* WebKitLibraries/DownlevelFrameworkStubs/130000/AppleJPEGXL.framework/AppleJPEGXL.tbd:

Canonical link: https://commits.webkit.org/265440@main
